File: 18-四数之和.py
class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []

        result = []
        nums.sort()
        t2 = 1
        
        while t2 < len(nums) - 2:
            t3 = t2 + 1
            while t3 < len(nums) - 1:
                t1 = 0
                t4 = len(nums) - 1
                while t1 < t2 and t3 < t4:
                    value = nums[t1] + nums[t2] + nums[t3] + nums[t4]
                    if value == target:
                        result.append([nums[t1], nums[t2], nums[t3], nums[t4]])
                        t1 += 1
                        t4 -= 1
                    elif value > target:
                        t4 -= 1
                    else:
                        t1 += 1
                t3 += 1
            t2 += 1

        result.sort()
        i = 0
        while i < len(result) - 1:
            if result[i] == result[i + 1]:
                del result[i + 1]
            else:
                i += 1

        return result

File: 18-四数之和.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        if(nums.size() < 4)
            return vector<vector<int>> {};

        vector<vector<int>> result;
        int t1 = 0, t2, t3, t4, value;
        sort(nums.begin(), nums.end());

        while(t1 < nums.size() - 3){
            t2 = t1 + 1;

            while(t2 < nums.size() - 2){
                t3 = t2 + 1;
                t4 = nums.size() - 1;

                while(t3 < t4){
                    value = nums[t1] + nums[t2] + nums[t3] + nums[t4];

                    if(value == target){
                        result.push_back(vector<int> {nums[t1], nums[t2], nums[t3], nums[t4]});
                        while(nums[++t3] == nums[t3 - 1] && nums[--t4] == nums[t4 + 1] && t3 < t4);
                    }
                    
                    else if(value > target)
                        t4 -= 1;
                    
                    else
                        t3 += 1;
                }

                while(nums[++t2] == nums[t2 - 1] && t2 < nums.size() - 2);

            }

            while(nums[++t1] == nums[t1 - 1] && t1 < nums.size() - 3);

        }

        return result;
    }
};
